File: Algorithms/src/3517. Smallest Palindromic Rearrangement I/Code/solution.java
class Solution {
    public String smallestPalindrome(String s) {

        // Store frequency of every lowercase letter
        int[] freq = new int[26];

        // Count character frequencies
        for (char c : s.toCharArray())
            freq[c - 'a']++;

        StringBuilder left = new StringBuilder();
        String middle = "";

        // Build the left half and find the middle character
        for (int i = 0; i < 26; i++) {

            // Add half of the occurrences to the left side
            for (int j = 0; j < freq[i] / 2; j++)
                left.append((char) ('a' + i));

            // Odd frequency character becomes the center
            if (freq[i] % 2 == 1)
                middle = String.valueOf((char) ('a' + i));
        }

        // Right half is the reverse of the left half
        String right = new StringBuilder(left).reverse().toString();

        // Return the complete palindrome
        return left.toString() + middle + right;
    }
}

File: Algorithms/src/3517. Smallest Palindromic Rearrangement I/Code/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var smallestPalindrome = function (s) {
  // Store frequency of every lowercase letter
  const freq = new Array(26).fill(0);

  // Count character frequencies
  for (const ch of s) freq[ch.charCodeAt(0) - 97]++;

  let left = "";
  let middle = "";

  // Build the left half and find the middle character
  for (let i = 0; i < 26; i++) {
    // Add half of the occurrences to the left side
    left += String.fromCharCode(97 + i).repeat(Math.floor(freq[i] / 2));

    // Odd frequency character becomes the center
    if (freq[i] % 2) middle = String.fromCharCode(97 + i);
  }

  // Right half is the reverse of the left half
  const right = left.split("").reverse().join("");

  // Return the complete palindrome
  return left + middle + right;
};
